lnst/Recipes/ENRT/ConfigMixins/FirewallMixin.py

```diff
@@ -0,0 +1,197 @@
+from lnst.Recipes.ENRT.ConfigMixins import BaseSubConfigMixin
+from tempfile import mkstemp
+from os import close, write, unlink
+import logging
+import copy
+
+class FirewallMixin(BaseSubConfigMixin):
+    """
+    A config mixin to apply custom firewall rulesets on hosts.
+    Do not inherit directly, use one of the derived classes below instead.
+    """
+
+    @property
+    def firewall_rulesets(self):
+        """
+        This property holds a dictionary of firewall rulesets in textual
+        representation, indexed by the host it should be applied to. A typical
+        use is:
+
+        { self.matched.host1: <host1 ruleset>,
+          self.matched.host2: <host2 ruleset> }
+
+        The rulesets will be applied by a derived class's _apply_ruleset()
+        method, i.e. typically fed into 'nft -f' or 'iptables-restore'.
+        """
+        return {}
+
+    def _apply_ruleset(self, host, path):
+        raise NotImplementedError("Method _apply_ruleset must be defined by a child class.")
+
+    def _dump_ruleset(self, host, config):
+        raise NotImplementedError("Method _dump_ruleset must be defined by a child class.")
+
+    def _push_ruleset(self, host, ruleset):
+        """
+        Push `ruleset` into a temporary file on `host`, returning its path.
+        """
+        fd, name = mkstemp(prefix="FirewallMixin", text=True)
+        write(fd, ruleset.encode('utf-8'))
+        close(fd)
+        remote_path = host.copy_file_to_machine(name)
+        unlink(name)
+        return remote_path
+
+    def apply_sub_configuration(self, config):
+        super().apply_sub_configuration(config)
+
+        if not config.firewall_rulesets:
+            config.firewall_rulesets = self.firewall_rulesets
+
+        self.stored_rulesets = {}
+        for host, ruleset in config.firewall_rulesets.items():
+            self.stored_rulesets[host] = self._dump_ruleset(host, config)
+
+            name = self._push_ruleset(host, ruleset)
+            self._apply_ruleset(host, name)
+            host.run('rm -f "%s"' % name)
+
+    def generate_sub_configuration_description(self, config):
+        desc = super().generate_sub_configuration_description(config)
+
+        for host, ruleset in config.firewall_rulesets.items():
+            desc.append("Firewall: ruleset with %d lines on host %s" % (len(ruleset.split("\n")), host))
+
+        return desc
+
+    def remove_sub_configuration(self, config):
+        for host, ruleset in self.stored_rulesets.items():
+            # XXX: this may spam debug log, but counter values may be worth logging
+            logging.debug("Dropping Firewall ruleset on host %s:\n%s\n" % (host, self._dump_ruleset(host, config)))
+            name = self._push_ruleset(host, ruleset)
+            self._apply_ruleset(host, name)
+            host.run('rm -f "%s"' % name)
+
+        return super().remove_sub_configuration(config)
+
+class BulkFirewallMixin(FirewallMixin):
```

Comment:
i feel like this class - the generator method and the subconfig generator should be part of the base `FirewallMixin` class... i mean, without this, the subclasses don't really do anything on their own, and with this addition, they would still by default do the default "nothing" operation.

But I'm not sure if you had some more specific reason to separate this into its own class?

Reply:
Merged both and they integrate nicely by making the generator method return '[self.firewall_rulesets]' by default.